node/src/exchange/server.rs

```diff
@@ -1,52 +1,396 @@
+
+    fn send_response(&mut self, channel: Self::Channel, response: ResponseType) {
+        // response was prepared specifically for the request, we can drop it
+        // in case of error we'll get Event::InboundFailure
+        let _ = self.send_response(channel, response);
+    }
 }
 
-impl<S> ExchangeServerHandler<S>
+impl<S, R> ExchangeServerHandler<S, R>
 where
     S: Store + 'static,
+    R: ResponseSender,
 {
     pub(super) fn new(store: Arc<S>) -> Self {
-        ExchangeServerHandler { _store: store }
+        ExchangeServerHandler {
+            store,
+            store_jobs: FuturesUnordered::new(),
+        }
     }
 
-    #[instrument(level = "trace", skip(self, _respond_to))]
-    pub(super) fn on_request_received(
+    #[instrument(level = "trace", skip(self, response_channel))]
+    pub(super) fn on_request_received<ID>(
         &mut self,
         peer: PeerId,
-        request_id: RequestId,
+        request_id: ID,
         request: HeaderRequest,
-        _respond_to: ResponseChannel<Vec<HeaderResponse>>,
-    ) {
-        // TODO
+        response_channel: R::Channel,
+    ) where
+        ID: Display + Debug,
+    {
+        let Some((amount, data)) = self.parse_request(request) else {
+            self.store_jobs
+                .push(handle_invalid_request(response_channel).boxed());
+            return;
+        };
+
+        let store_job = match data {
+            header_request::Data::Origin(0) => {
+                handle_request_current_head(self.store.clone(), response_channel).boxed()
+            }
+            header_request::Data::Origin(height) => {
+                handle_request_by_height(self.store.clone(), response_channel, height, amount)
+                    .boxed()
+            }
+            header_request::Data::Hash(hash) => {
+                handle_request_by_hash(self.store.clone(), response_channel, hash).boxed()
+            }
+        };
+
+        self.store_jobs.push(store_job);
     }
 
-    #[instrument(level = "trace", skip(self))]
     pub(super) fn on_response_sent(&mut self, peer: PeerId, request_id: RequestId) {
-        // TODO
+        trace!("response_sent; request_id: {request_id}, peer: {peer}");
     }
 
-    #[instrument(level = "trace", skip(self))]
     pub(super) fn on_failure(
         &mut self,
         peer: PeerId,
         request_id: RequestId,
         error: InboundFailure,
     ) {
-        // TODO
+        // TODO: cancel job if libp2p already failed it?
+        trace!("on_failure; request_id: {request_id}, peer: {peer}, error: {error:?}");
+    }
+
+    fn parse_request(&self, request: HeaderRequest) -> Option<(u64, header_request::Data)> {
+        if !request.is_valid() {
+            return None;
+        }
+
+        let HeaderRequest {
+            amount,
+            data: Some(data),
+        } = request
+        else {
+            return None;
+        };
+
+        Some((amount, data))
+    }
```

Comment:
This is a utility, so have it as standalone function

Reply:
done 👍